core/src/main/java/org/infinispan/marshall/core/GlobalMarshaller.java

```diff
@@ -370,8 +414,189 @@ private void writeNonNullableObject(Object obj, BytesObjectOutput out) throws IO
       }
    }
 
+   private void writeArray(Class<?> clazz, Object array, BytesObjectOutput out) throws IOException {
+      out.writeByte(ID_ARRAY);
+      Class<?> componentType = clazz.getComponentType();
+
+      int length = Array.getLength(array);
```

Comment:
Haven't micro-benchmarked that, and it's native code. I guess that compilers can optimize this pretty well.